packages/api/src/clickhouse/index.ts

```diff
@@ -904,6 +912,588 @@ export const getMetricsChart = async ({
+  const serializer = new SQLSerializer(propertyTypeMappingsModel);
+
+  const label = SqlString.escape(`${aggFn}(${field})`);
+
+  const selectClause = [
+    isCountFn
+      ? 'toFloat64(count()) as data'
+      : aggFn === AggFn.Sum
+      ? `sum(${selectField}) as data`
+      : aggFn === AggFn.Avg
+      ? `avg(${selectField}) as data`
+      : aggFn === AggFn.Max
+      ? `max(${selectField}) as data`
+      : aggFn === AggFn.Min
+      ? `min(${selectField}) as data`
+      : aggFn === AggFn.CountDistinct
+      ? `count(distinct ${selectField}) as data`
+      : `quantile(${
+          aggFn === AggFn.P50
+            ? '0.5'
+            : aggFn === AggFn.P90
+            ? '0.90'
+            : aggFn === AggFn.P95
+            ? '0.95'
+            : '0.99'
+        })(${selectField}) as data`,
+    granularity != null
+      ? `toUnixTimestamp(toStartOfInterval(timestamp, INTERVAL ${granularity})) as ts_bucket`
+      : "'0' as ts_bucket",
+    groupByField ? `${groupByField} as group` : `'' as group`, // FIXME: should we fallback to use aggFn as group
+    `${label} as label`,
+  ].join(',');
+
+  const groupByClause = `ts_bucket ${groupByField ? `, ${groupByField}` : ''}`;
+
+  const query = SqlString.format(
+    `
+      SELECT ?
+      FROM ??
+      WHERE ? AND (?) ? ?
+      GROUP BY ?
+      ORDER BY ts_bucket ASC
+      ${
+        granularity != null
+          ? `WITH FILL
+        FROM toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        TO toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        STEP ?`
+          : ''
+      }${
+      sortOrder === 'asc' || sortOrder === 'desc' ? `, data ${sortOrder}` : ''
+    }
+    `,
+    [
+      SqlString.raw(selectClause),
+      tableName,
+      buildTeamLogStreamWhereCondition(tableVersion, teamId),
+      SqlString.raw(whereClause),
+      SqlString.raw(
+        !isCountFn && field != null
+          ? ` AND (${await serializer.isNotNull(field, false)})`
+          : '',
+      ),
+      SqlString.raw(
+        hasGroupBy
+          ? ` AND (${await serializer.isNotNull(groupBy, false)})`
+          : '',
+      ),
+      SqlString.raw(groupByClause),
+      ...(granularity != null
+        ? [
+            startTime / 1000,
+            granularity,
+            endTime / 1000,
+            granularity,
+            ms(granularity) / 1000,
+          ]
+        : []),
+    ],
+  );
+
+  return {
+    query,
+    hasGroupBy,
+  };
+};
+
+export const queryMultiSeriesChart = async ({
+  maxNumGroups,
+  tableVersion,
+  teamId,
+  seriesReturnType = 'column',
+  queries,
+}: {
+  maxNumGroups: number;
+  tableVersion: number | undefined;
+  teamId: string;
+  seriesReturnType?: 'ratio' | 'column';
+  queries: { query: string; hasGroupBy: boolean }[];
+}) => {
+  // For now only supports same-table series with the same groupBy
+
+  const seriesCTEs = SqlString.raw(
+    'WITH ' + queries.map((q, i) => `series_${i} AS (${q.query})`).join(',\n'),
```

Comment:
style suggestion: I'd move 'WITH' to the final query so its clear that we are building a bunch of CTEs